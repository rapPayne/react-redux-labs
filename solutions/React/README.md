# React Solutions

These are the solutions for React (not React Native) courses.

They will:
- be upgraded to React 19
- use TypeScript (?)
- use Zustand as the application state management
- use useActionState for forms
- use the use hook for async fetches
- use custom hooks for common tasks like fetching things
- use the CSS-per-component pattern

## Redux notes
The React 5-day class teaches Redux also. The 3-day version has only a "Redux in a nutshell" lecture/lab with and a bunch of pre-written Redux code. 
- [Redux class lab solutions](Redux)
- [Redux in a nutshell solution](030-Redux-in-a-nutshell)
