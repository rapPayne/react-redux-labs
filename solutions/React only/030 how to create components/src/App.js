import React, { useState, useEffect } from 'react';
import { store } from './store/store';
import { actions } from './store/actions';
import { LandingPage } from './LandingPage';
import { FilmDetails } from './FilmDetails';
import { PickSeats } from './PickSeats';
import { Checkout } from './Checkout';
import { NotFound } from './NotFound';
import { Login } from './authentication/Login';
import { Logout } from './authentication/Logout';
import { Account } from './authentication/Account';

function App() {
  const [state, setState] = useState(store.getState());
  useEffect(() => {
    store.subscribe(() => setState({ ...store.getState() }));
    store.dispatch(actions.fetchInitialData());
  }, []);
  console.log(store.getState());
  return (
    <div className="mdl-layout mdl-js-layout mdl-layout--fixed-header">
      <header className="mdl-layout__header">
        <div className="mdl-layout__header-row">
          <a href="/" className="mdl-layout-title">Dinner and a Movie</a>
          <nav className="mdl-navigation mdl-layout--large-screen-only">
            <a href="/account" className="mdl-layout__tab">My account</a>
            <a href="/logout" className="mdl-layout__tab">logout</a>
            <a href="/checkout" className="mdl-layout__tab"><i className="material-icons">shopping_cart</i></a>
            <a href="/login" className="mdl-layout__tab">Login</a>
            <a href="/register" className="mdl-layout__tab">Register</a>
          </nav>
        </div>
      </header>
      <div className="mdl-layout__drawer">
        <a href="/" className="mdl-layout-title">Dinner and a Movie</a>
        <nav className="mdl-navigation">
          <a href="/account" className="mdl-layout__link">My account</a>
          <a href="/logout" className="mdl-layout__link">logout</a>
          <a href="/checkout" className="mdl-layout__link"><i className="material-icons">shopping_cart</i></a>
          <a href="/login" className="mdl-layout__link">Login</a>
          <a href="/register" className="mdl-layout__link">Register</a>
        </nav>
      </div>
      <main className="mdl-layout__content">
        <Logout />
      </main>
      <footer>
      </footer>
    </div>
  );
}

export default App;
