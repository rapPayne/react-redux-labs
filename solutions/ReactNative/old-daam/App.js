import React, { useState, useEffect } from 'react';
import { store } from './store/store';
import { StyleSheet, Text, View } from 'react-native';
import { Landing } from './Landing';


export default function App() {
  const [state, setState] = useState(store.getState());
  useEffect(() => {
    store.subscribe(() => setState(store.getState()));
    store.dispatch({type:"FETCH_FILMS"})
  }, []);
  return (
    <View style={styles.container}>
      <Landing {...state}></Landing>
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#fff',
    alignItems: 'center',
    justifyContent: 'center',
  },
});
