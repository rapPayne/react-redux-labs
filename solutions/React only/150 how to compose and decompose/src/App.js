import React, { useState, useEffect } from 'react';
import 'material-design-lite/dist/material.min.css';
import 'material-design-lite/dist/material.purple-indigo.min.css';
import './App.css';
import './helpers/Currency';
import './helpers/Date';
import 'material-design-lite/material';
import { store } from './store/store';
import { actions } from './store/actions';
import { LandingPage } from './LandingPage';
import { FilmDetails } from './FilmDetails';
import { PickSeats } from './PickSeats';
import { Checkout } from './Checkout';
import { NotFound } from './NotFound';
import { Login } from './authentication/Login';
import { Logout } from './authentication/Logout';
import { Account } from './authentication/Account';

function App() {
  const [state, setState] = useState(store.getState());
  useEffect(() => {
    store.subscribe(() => setState({ ...store.getState() }));
    store.dispatch(actions.fetchInitialData());
  }, []);
  console.log(store.getState());
  console.log(1234.567.toCurrency());
  return (
    <div className="mdl-layout mdl-js-layout mdl-layout--fixed-header">
      <header className="mdl-layout__header">
        <div className="mdl-layout__header-row">
          <a href="/" style={{ ...styles.navlink, ...styles.topMenuNavLink }} className="mdl-layout-title">Dinner and a Movie</a>
          <nav className="mdl-navigation mdl-layout--large-screen-only">
            {state.user ?
              <>
                <a href="/account" className="mdl-layout__tab">My account</a>
                <a href="/logout" className="mdl-layout__tab">logout</a>
                <a href="/checkout" className="mdl-layout__tab"><i className="material-icons">shopping_cart</i></a>
              </>
              :
              <>
                <a href="/login" className="mdl-layout__tab">Login</a>
                <a href="/register" className="mdl-layout__tab">Register</a>
              </>}
          </nav>
        </div>
      </header>
      <div className="mdl-layout__drawer">
        <a href="/" style={{ ...styles.drawerNavLink, ...styles.navlink }} className="mdl-layout-title">Dinner and a Movie</a>
        <nav className="mdl-navigation">
          {state.user ?
            <>
              <a href="/account" className="mdl-layout__link">My account</a>
              <a href="/logout" className="mdl-layout__link">logout</a>
              <a href="/checkout" className="mdl-layout__link"><i className="material-icons">shopping_cart</i></a>
            </>
            :
            <>
              <a href="/login" className="mdl-layout__link">Login</a>
              <a href="/register" className="mdl-layout__link">Register</a>
            </>}
        </nav>
      </div>
      <main className="mdl-layout__content">
        <FilmDetails />
      </main>
      <footer>
      </footer>
    </div>
  );
}

const styles = {
  navlink: {
    padding: '10px',
    textTransform: 'uppercase',
    textDecoration: 'none',
  },
  drawerNavLink: {
    color: '#424242',
  },
  topMenuNavLink: {
    color: 'white',
  },
}

export default App;
