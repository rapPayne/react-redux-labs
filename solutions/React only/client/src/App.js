import React, { useEffect, useState } from 'react';
import { BrowserRouter, Route, Switch, Link } from 'react-router-dom';
import 'material-design-lite/dist/material.min.css';
import 'material-design-lite/dist/material.purple-indigo.min.css';
import 'material-design-lite/material';
import './helpers/Date';
import './helpers/Currency';
import './App.css';
import { Account } from './authentication/Account';
import { Checkout } from './Checkout';
import { FilmDetails } from './FilmDetails';
import { LandingPage } from './LandingPage';
import { Login } from './authentication/Login';
import { Logout } from './authentication/Logout';
import { NotFound } from './NotFound';
import { PickSeats } from './PickSeats';
import { actions } from './store/actions';
import { store } from './store/store';

function App() {
  const [state, setState] = useState(store.getState());
  useEffect(() => {
    store.subscribe(() => setState({ ...store.getState() }));
    store.dispatch(actions.fetchInitialData());
  }, []);

  return (
    <BrowserRouter>
      <div className="mdl-layout mdl-js-layout mdl-layout--fixed-header">
        <header className="mdl-layout__header">
          <div className="mdl-layout__header-row">
            <Link to="/" style={{ ...styles.navlink, ...styles.topMenuNavLink }} className="mdl-layout-title">Dinner and a Movie</Link>
            <nav className="mdl-navigation mdl-layout--large-screen-only">
              {state.user ?
                <>
                  <Link to="/account" className="mdl-layout__tab">My account</Link>
                  <Link to="/logout" className="mdl-layout__tab">logout</Link>
                  <Link to="/checkout" className="mdl-layout__tab"><i className="material-icons">shopping_cart</i></Link>
                </>
                :
                <>
                  <Link to="/login" className="mdl-layout__tab">Login</Link>
                  <Link to="/register" className="mdl-layout__tab">Register</Link>
                </>
              }
            </nav>
          </div>
        </header>
        <div className="mdl-layout__drawer">
          <Link to="/" style={{ ...styles.drawerNavLink, ...styles.navlink }} className="mdl-layout-title">Dinner and a Movie</Link>
          <nav className="mdl-navigation">
            {state.user ?
              <>
                <Link to="/account" className="mdl-layout__link">My account</Link>
                <Link to="/logout" className="mdl-layout__link">logout</Link>
                <Link to="/checkout" className="mdl-layout__link"><i className="material-icons">shopping_cart</i></Link>
              </>
              :
              <>
                <Link to="/login" className="mdl-layout__link">Login</Link>
                <Link to="/register" className="mdl-layout__link">Register</Link>
              </>
            }
          </nav>
        </div>
        <main className="mdl-layout__content">
          <Switch>
            <Route exact path="/" render={_ => <LandingPage {...state} />} />
            <Route exact path="/account" render={_ => <Account {...state} />} />
            <Route exact path="/login" render={_ => <Login />} />
            <Route exact path="/logout" render={_ => <Logout />} />
            <Route exact path="/register" render={_ => <Account />} />
            <Route exact path="/checkout" render={_ => <Checkout {...state} />} />
            <Route exact path="/pickseats/:showingId" render={_ => <PickSeats {...state} />} />
            <Route exact path="/filmdetails/:filmId" render={_ => <FilmDetails {...state} />} />
            <Route render={_ => <NotFound />} />
          </Switch>
        </main>
        <footer>
        </footer>
      </div>
    </BrowserRouter>
  );
}

const styles = {
  navlink: {
    padding: '10px',
    textTransform: 'uppercase',
    textDecoration: 'none',
  },
  drawerNavLink: {
    color: '#424242',
  },
  topMenuNavLink: {
    color: 'white',
  },
}
export default App;
