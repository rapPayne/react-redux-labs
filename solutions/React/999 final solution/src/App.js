import { useEffect, useMemo } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { BrowserRouter, Route, Routes, Link } from 'react-router-dom';
import 'material-design-lite/dist/material.min.css';
import 'material-design-lite/dist/material.purple-indigo.min.css';
import 'material-design-lite/material';
import './helpers/Date';
import './helpers/Currency';
import './App.css';
import { Account } from './authentication/Account';
import { Checkout } from './Checkout';
import { FilmDetails } from './FilmDetails';
import { LandingPage } from './LandingPage';
import { Login } from './authentication/Login';
import { Logout } from './authentication/Logout';
import { NotFound } from './NotFound';
import { PickSeats } from './PickSeats';
import { actions } from './store/actions';

function App() {
  const state = useSelector(state => state);
  const user = useSelector(state => state.user);
  const dispatch = useDispatch();
  useEffect(() => {
    dispatch(actions.fetchInitialData());
  }, [dispatch]);

  return (
    <BrowserRouter>
      <div className="mdl-layout mdl-js-layout mdl-layout--fixed-header">
        <header className="mdl-layout__header">
          <div className="mdl-layout__header-row">
            <Link to="/" style={{ ...styles.navlink, ...styles.topMenuNavLink }} className="mdl-layout-title">Dinner and a Movie</Link>
            <nav className="mdl-navigation mdl-layout--large-screen-only">
              {user ?
                <>
                  <Link to="/account" className="mdl-layout__tab">My account</Link>
                  <Link to="/logout" className="mdl-layout__tab">logout</Link>
                  <Link to="/checkout" className="mdl-layout__tab"><i className="material-icons">shopping_cart</i></Link>
                </>
                :
                <>
                  <Link to="/login" className="mdl-layout__tab">Login</Link>
                  <Link to="/register" className="mdl-layout__tab">Register</Link>
                </>
              }
            </nav>
          </div>
        </header>
        <div className="mdl-layout__drawer">
          <Link to="/" style={{ ...styles.drawerNavLink, ...styles.navlink }} className="mdl-layout-title">Dinner and a Movie</Link>
          <nav className="mdl-navigation">
            {user ?
              <>
                <Link to="/account" className="mdl-layout__link">My account</Link>
                <Link to="/logout" className="mdl-layout__link">logout</Link>
                <Link to="/checkout" className="mdl-layout__link"><i className="material-icons">shopping_cart</i></Link>
              </>
              :
              <>
                <Link to="/login" className="mdl-layout__link">Login</Link>
                <Link to="/register" className="mdl-layout__link">Register</Link>
              </>
            }
          </nav>
        </div>
        <main className="mdl-layout__content">
          <Routes>
            <Route path="/" element={<LandingPage />} />
            <Route path="/account" element={<Account {...state} />} />
            <Route path="/login" element={<Login />} />
            <Route path="/logout" element={<Logout />} />
            <Route path="/register" element={<Account />} />
            <Route path="/checkout" element={<Checkout {...state} />} />
            <Route path="/pickseats/:showingId" element={<PickSeats {...state} />} />
            <Route path="/film/:filmId" element={<FilmDetails {...state} />} />
            <Route path="*" element={<NotFound />} />
          </Routes>
        </main>
        <footer>
        </footer>
      </div>
    </BrowserRouter>
  );
}

const styles = {
  navlink: {
    padding: '10px',
    textTransform: 'uppercase',
    textDecoration: 'none',
  },
  drawerNavLink: {
    color: '#424242',
  },
  topMenuNavLink: {
    color: 'white',
  },
}
export default App;
