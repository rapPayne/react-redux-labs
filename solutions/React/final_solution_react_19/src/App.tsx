
import { ReactElement, Suspense, use, useEffect, useTransition } from 'react'
import './App.css'
import { BrowserRouter, Route, Routes } from 'react-router-dom';
import { LandingPage } from './components/scenes/LandingPage.tsx';
import { Account } from './components/authentication/Account.tsx';
import { Login } from './components/authentication/Login.tsx';
//import { Login } from './components/authentication/LoginActionState.tsx';
//import { Login } from './components/authentication/LoginTestServerAction.tsx';
import { Logout } from './components/authentication/Logout.tsx';
import { Checkout } from './components/scenes/Checkout.tsx';
import { PickSeats } from './components/scenes/PickSeats.tsx';
import { FilmDetails } from './components/scenes/FilmDetails.tsx';
import { NotFound } from './components/scenes/NotFound.tsx';
import { State, useStore } from './store/useStore.ts';
import { Convert as ConvertFilm, Film } from './types/Film.ts';
import { Convert as ConvertShowing } from './types/Showing.ts';
import { Convert as ConvertTheater } from './types/Theater.ts';
import { FilmList } from './components/FilmListWithUse.tsx';
// import { FilmList } from './components/FilmListWithFormActionAttr.tsx';
import { User } from './types/User.ts';
import { ResponsiveNavBar } from './components/ResponsiveNavBar.tsx';
import { Toaster } from 'react-hot-toast';

const filmsPromise =
  fetch("http://localhost:3008/films")
    .then(res => res.text())
    .then(json => ConvertFilm.toFilm(json));

// fetch("http://localhost:3008/films")
//   .then(res => res.text())
//   .then(json => ConvertFilm.toFilm(json))
//   .then(films => store.setFilms(films))
//   .catch(console.error);

export function App(): ReactElement {
  const store: State = useStore();
  const user: User | undefined = store.user;
  let films = store.films;
  if (!films) {
    films = use(filmsPromise)
    store.setFilms(films)
  }

  useEffect(() => {
    // Load initial data here
    // fetch("http://localhost:3008/films")
    //   .then(res => res.text())
    //   .then(json => ConvertFilm.toFilm(json))
    //   .then(films => store.setFilms(films))
    //   .catch(console.error);

    fetch("http://localhost:3008/theaters")
      .then(res => res.text())
      .then(json => ConvertTheater.toTheater(json))
      .then(theaters => store.setTheaters(theaters))
      .catch(console.error);

    //TODO: This is super inefficient. Should only fetch showings we need.
    fetch("http://localhost:3008/showings")
      .then(res => res.text())
      .then(json => ConvertShowing.toShowing(json))
      .then(showings => store.setShowings(showings))
      .catch(console.error);

  }, []);
  return (
    <BrowserRouter>
      <div className="mdl-layout mdl-js-layout mdl-layout--fixed-header">
        <header>
          <Toaster position="top-right" reverseOrder={true} />
          <ResponsiveNavBar user={user} />
        </header>
        <main className="mdl-layout__content">
          <Suspense fallback={<h1>Loading. Please wait ...</h1>}>
            {/* <FilmList filmsPromise={filmsPromise} /> */}
            <Routes>
              <Route path="/" element={<LandingPage />} />
              <Route path="/account" element={<Account />} />
              <Route path="/login" element={<Login />} />
              <Route path="/logout" element={<Logout />} />
              <Route path="/register" element={<Account />} />
              <Route path="/checkout" element={<Checkout cart={store.cart} user={store.user} />} />
              <Route path="/pickseats/:showingId" element={<PickSeats />} />
              <Route path="/film/:filmId" element={<FilmDetails />} />
              <Route path="*" element={<NotFound />} />
            </Routes>
          </Suspense>
        </main>
        <footer>
        </footer>
      </div>
    </BrowserRouter>
  );
}

export default App