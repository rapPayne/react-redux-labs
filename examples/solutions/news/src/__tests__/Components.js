import React from 'react';
import ReactDOM from 'react-dom';
import { shallow } from 'enzyme';
import { Dogs } from '../Dogs';
import App from '../App';

import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
configure({ adapter: new Adapter() });

beforeEach(() => {
  Math.random = jest.fn(() => 23)
});

describe("dogs", () => {
  it('Renders without crashing', () => {
    Math.random = jest.fn(() => 23)
    const rand = Math.random();
    const dogsComponent = shallow(<Dogs number={rand} />);
    expect(dogsComponent).toMatchSnapshot();
  }); 

  it('have a random number that is not so random', () => {
    const random = Math.random();
    expect(random).toEqual(23); 
  })
});

class TestWrapper extends React.Component {
  render() {
    return this.props.children;
  }
}

