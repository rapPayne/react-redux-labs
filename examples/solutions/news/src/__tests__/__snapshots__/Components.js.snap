// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dogs Renders without crashing 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Dogs
    number={23}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <h1>
          I love dogs!
        </h1>,
        <p>
          23
        </p>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "I love dogs!",
        },
        "ref": null,
        "rendered": "I love dogs!",
        "type": "h1",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": 23,
        },
        "ref": null,
        "rendered": 23,
        "type": "p",
      },
    ],
    "type": Symbol(react.fragment),
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <h1>
            I love dogs!
          </h1>,
          <p>
            23
          </p>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "I love dogs!",
          },
          "ref": null,
          "rendered": "I love dogs!",
          "type": "h1",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": 23,
          },
          "ref": null,
          "rendered": 23,
          "type": "p",
        },
      ],
      "type": Symbol(react.fragment),
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
